src/main/scala/is/hail/variant/VariantSampleMatrix.scala

```diff
@@ -1609,6 +1609,47 @@ class VariantSampleMatrix[RPK, RK, T >: Null](val hc: HailContext, val metadata:
     (t, f2)
   }
 
+  def reorderSamples(newIds: java.util.ArrayList[Annotation]): VariantSampleMatrix[RPK, RK, T] =
+    reorderSamples(newIds.asScala.toArray)
+
+  def reorderSamples(newIds: Array[Annotation]): VariantSampleMatrix[RPK, RK, T] = {
+    requireUniqueSamples("reorder_samples")
+
+    val oldOrder = sampleIds.zipWithIndex.toMap
+    val newOrder = newIds.zipWithIndex.toMap
+
+    val mapping = oldOrder.outerJoin(newOrder).map { case (s, (oldIdx, newIdx)) =>
+      (oldIdx, newIdx) match {
+        case (Some(i), Some(j)) => (i, j)
+        case (None, Some(j)) => fatal(s"`$s' is not in dataset.")
+        case (Some(i), None) => fatal(s"dataset sample id `$s' not found in new id list.")
+        case (None, None) => fatal(s"This case should never happen!")
+      }
+    }
+
+    val newIndices = sampleIds.indices.map(mapping)
+
+    val newAnnotations = new Array[Annotation](nSamples)
+    sampleAnnotations.zipWithIndex.foreach { case (sa, idx) =>
+      newAnnotations(newIndices(idx)) = sa
+    }
+
+    val nSamplesLocal = nSamples
+    val localtct = tct
+
+    val reorderedRdd = rdd.mapPartitions ({ it =>
+        it.map { case (v, (va, gs)) =>
+          val reorderedGs = Array.fill[T](nSamplesLocal)(null)(localtct)
+          gs.zipWithIndex.foreach { case (g, i) =>
+            reorderedGs(newIndices(i)) = g
+          }
+          (v, (va, reorderedGs.toIterable))
```

Comment:
It's probably worth writing this as
```
var i = 0
while (gs.hasNext) {
  reorderedGs(newIndices(i)) = gs.next
  i += 1
}
```
considering it will execute for every variant in the dataset.